### PR TITLE
[fix][txn] Fix the transaction acknowledgement and send logic for  chunk message

### PR DESCRIPTION
**Pull request by liangyepianzhou**
Master https://github.com/apache/pulsar-client-go/issues/1060
### Motivation
1. For the chunk message, we only register the send operation once but end the send operation multiple times when receiving the send response. It will make the transaction can be committed before all the operations are completed.
2. When we use transaction ack for chunk messages, the provided transaction is ignored, resulting in the chunk message actually being acknowledged using the non-transactional ack method.
### Modifications
1. Only end the send operation when receive the last chunk message.
2. Add the check for the transaction when the massage is a chunk message.
### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -984,7 +984,7 @@ func (p *partitionProducer) failTimeoutMessages() {
 						sr.callback(nil, sr.msg, errSendTimeout)
 					})
 				}
-				if sr.transaction != nil {
+				if (sr.totalChunks <= 1 || sr.chunkID == sr.totalChunks-1) && sr.transaction != nil {
 					sr.transaction.endSendOrAckOp(nil)
```
Comment (gunli) [inline pulsar/producer_partition.go:988]: I think it should passed `errSendTimeout` to `endSendOrAckOp` to update the transcaction to `TxnError`
```go
func (txn *transaction) endSendOrAckOp(err error) {
	if err != nil {
		atomic.StoreInt32((*int32)(&txn.state), int32(TxnError))
	}
	if atomic.AddInt32(&txn.opsCount, -1) == 0 {
		//This means there are not pending send/ack requests
		txn.opsFlow <- true
	}
}
```

File: pulsar/producer_partition.go [hunk 2]
```diff
@@ -979,14 +979,11 @@ func (p *partitionProducer) failTimeoutMessages() {
 						WithField("properties", sr.msg.Properties)
 				}
 
-				if sr.callback != nil {
+				if sr.totalChunks <= 1 || sr.chunkID == sr.totalChunks-1 {
 					sr.callbackOnce.Do(func() {
 						sr.callback(nil, sr.msg, errSendTimeout)
```
Comment (gunli) [inline pulsar/producer_partition.go:984]: run callback without checking if it is nil

**Comment by gunli** [comment]
@liangyepianzhou The bug of producer I have fixed in #1071

**Comment by liangyepianzhou** [comment]
> @liangyepianzhou The bug of producer I have fixed in #1071

Maybe you can merge the master after this PR is merged?

**Comment by gunli** [comment]
Or you can just fix the consumer, I have refactored the producer, that modified a lot of code.

**Comment by gunli** [comment]
You are welcome. So, if you are free, would you please help to review my PR #1071.

**Review by RobertIndie — APPROVED** [review]

**Pull request merged**